DigitalLearningSolutions.Data/Services/CourseDelegatesDownloadFileService.cs

```diff
@@ -0,0 +1,177 @@
+    {
+        private readonly ICourseAdminFieldsService courseAdminFieldsService;
+        private readonly ICourseDelegatesDataService courseDelegatesDataService;
+        private readonly ICourseService courseService;
+        private readonly ICentreCustomPromptsService customPromptsService;
+
+        public CourseDelegatesDownloadFileService(
+            ICourseService courseService,
+            ICourseDelegatesDataService courseDelegatesDataService,
+            ICourseAdminFieldsService courseAdminFieldsService,
+            ICentreCustomPromptsService customPromptsService
+        )
+        {
+            this.courseService = courseService;
+            this.courseDelegatesDataService = courseDelegatesDataService;
+            this.courseAdminFieldsService = courseAdminFieldsService;
+            this.customPromptsService = customPromptsService;
+        }
+
+        public byte[] GetCourseDelegateDownloadFileForCourse(int customisationId, int centreId, int? adminCategoryId)
+        {
+            using var workbook = new XLWorkbook();
+
+            PopulateCourseDelegatesSheetForCourse(workbook, customisationId, centreId, adminCategoryId);
+
+            using var stream = new MemoryStream();
+            workbook.SaveAs(stream);
+            return stream.ToArray();
+        }
+
+        private void PopulateCourseDelegatesSheetForCourse(
+            IXLWorkbook workbook,
+            int customisationId,
+            int centreId,
+            int? adminCategoryId
+        )
+        {
+            var course = courseService.GetCourseDetailsFilteredByCategory(customisationId, centreId, adminCategoryId)!;
+            var adminFields = courseAdminFieldsService.GetCustomPromptsForCourse(customisationId);
+
+            var customRegistrationPrompts = customPromptsService.GetCustomPromptsForCentreByCentreId(centreId);
+
+            var courseDelegates = courseDelegatesDataService.GetDelegatesOnCourseForExport(customisationId, centreId);
+
+            var dataTable = new DataTable();
+
+            SetUpDataTableColumns(customRegistrationPrompts, adminFields, dataTable);
+
+            foreach (var courseDelegate in courseDelegates.OrderBy(x => x.LastName))
+            {
+                AddDelegateToDataTable(dataTable, courseDelegate, customRegistrationPrompts, adminFields);
+            }
+
+            ClosedXmlHelper.AddSheetToWorkbook(
+                workbook,
+                $"Course {course.CustomisationId}",
+                dataTable.AsEnumerable(),
+                XLTableTheme.None
+            );
+        }
+
+        private static void SetUpDataTableColumns(
+            CentreCustomPrompts customRegistrationPrompts,
+            CourseAdminFields adminFields,
+            DataTable dataTable
+        )
+        {
+            dataTable.Columns.AddRange(
+                new[] { new DataColumn("Last name"), new DataColumn("First name"), new DataColumn("Email") }

```

Comment:
I think it would be good to define these magic strings in constants, more so because they are used more than once. Same for the others too.